test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/hex"
+	"fmt"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"io"
+	"strings"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	. "antrea.io/antrea/test/e2e/utils"
+)
+
+const interfaceNameLength = 5
+
+type vmInfo struct {
+	NodeName string
+	OS       string
+	IFName   string
+	IP       string
+	IFIndex  string // Used only for Windows
+}
+
+// TestBasic is the top-level test which contains some subtests for
+// basic test cases so they can share setup, teardown.
+func TestVMAgent(t *testing.T) {
+	skipIfHasWindowsNodes(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	t.Run("testExternalNodeBasic", func(t *testing.T) { testExternalNodeBasic(t, data, "vm-ns") })
+}
+
+func testExternalNodeBasic(t *testing.T, data *TestData, namespace string) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.winVMs, testOptions.linVMs)
+	nsReturned, _ := data.clientset.CoreV1().Namespaces().Get(context.TODO(), namespace, metav1.GetOptions{})
+	saReturned, _ := data.clientset.CoreV1().ServiceAccounts(namespace).Get(context.TODO(), "vm-agent", metav1.GetOptions{})
+	t.Logf("Using service account %s", saReturned.Name)
+	t.Logf("Using name space %s", nsReturned.Name)
+	var vmList []vmInfo
+	if testOptions.linVMs != "" {
+		vms := strings.Split(testOptions.linVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
+			tempVm := getVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	if testOptions.winVMs != "" {
+		vms := strings.Split(testOptions.winVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Windows VM, %s", vm)
+			tempVm := getWindowsVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+
+	t.Logf("=== TestSetup ===")
+	for _, vm := range vmList {
+		stopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalNode for %+v", vm)
+		_, err := createExternalNodeCRD(data, nsReturned.Name, vm.NodeName, vm.IFName, vm.IP)
+		require.Nil(t, err, "Failed to create ExternalNode")
+		startAntreaAgent(t, data, vm)
+	}
+
+	t.Logf("=== TestVerification ===")
+	for _, vm := range vmList {
```

Comment:
Use this format for verification?
```

	verifyExternalEntityExistence := func(eeName string, vmNodeName string, expectExists bool) {
		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
			t.Logf("Fetching ExternalEntity %s", eeName)
			_, err = data.crdClient.CrdV1alpha2().ExternalEntities(nsReturned.Name).Get(context.TODO(), eeName, metav1.GetOptions{})
			if err != nil && !errors.IsNotFound(err) {
				t.Logf("Failed to get ExternalEntity %s by ExternalNode %s: %v", eeName, vmNodeName, err)
				return false, err
			}

			if expectExists {
				if err != nil {
					return false, err
				}
				return true, nil
			}
			if err == nil {
				return true, nil
			}
			return false, nil

		}); err != nil {
			op := "created"
			if !expectExists {
				op = "deleted"
			}
			assert.NoError(t, err, "Failed to verify ExternalEntity %s %s by ExternalNode %s", eeName, op, vmNodeName)
		}
	}

	verifyExternalNodeRealization := func(vm vmInfo) {
		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
			t.Logf("Verify host interface configuration, VmInfo %+v", vm)
			exists, err := verifyInterfaceIsInOVS(t, data, vm)
			return exists, err
		}); err != nil {
			assert.NoError(t, err, "Failed to verify host interface in OVS, vmInfo %+v", vm)
			return
		}

		var tempVm vmInfo
		if vm.OS == "Windows" {
			tempVm = getWindowsVmInfo(t, data, vm.Nodename)
		} else {
			tempVm = getVmInfo(t, data, vm.Nodename)
		}
		assert.Exactly(t, vm.IfName, tempVm.IfName, "Failed to verify uplink interface, Expected %s, Got %s", vm.IfName, tempVm.IfName)
		assert.Exactly(t, vm.IP, tempVm.IP, "Failed to verify uplink IP, Expected %s, Got %s", vm.IP, tempVm.IP)
	}

	t.Logf("=== TestVerification ===")
	for _, vm := range vmList {
		hash := sha1.New()
		io.WriteString(hash, vm.IfName)
		hashedIfName := hex.EncodeToString(hash.Sum(nil))
		// Generate ExternalEntity name from ExternalNode
		eeName := vm.Nodename + "-" + hashedIfName[:interfaceNameLength]
		verifyExternalEntityExistence(eeName, vm.Nodename, true)
		verifyExternalNodeRealization(vm)
		err := data.crdClient.CrdV1alpha1().ExternalNodes(nsReturned.Name).Delete(context.TODO(), vm.Nodename, metav1.DeleteOptions{})
		require.Nil(t, err, "Failed to delete ExternalNode")
		verifyExternalEntityExistence(eeName, vm.Nodename, false)
	}
```

Besides, if the test time is long, we could think about use multiple go routines to run verification, each go routine verifies one VM, and we can use a `waitgroup` to ensure all the verify is completed on all VMs.

Reply:
I don't think the verification part takes time, so the testrun would finish quickly.

Reply:
Thanks for the suggestion, this looks much cleaner